File: mock-hss/tests/milenage.test.ts
import { describe, it, expect } from 'vitest';
import { computeOPc, generateVectorsWithRand, f5Star, f1Star, validateAuts, generateAuts } from '../src/milenage.js';

// 3GPP TS 35.207 Test Set 1
describe('MILENAGE Test Set 1 (TS 35.207)', () => {
  const ki = Buffer.from('465b5ce8b199b49faa5f0a2ee238a6bc', 'hex');
  const op = Buffer.from('cdc202d5123e20f62b6d676ac72cb318', 'hex');
  const rand = Buffer.from('23553cbe9637a89d218ae64dae47bf35', 'hex');
  const sqn = Buffer.from('ff9bb4d0b607', 'hex');
  const amf = Buffer.from('b9b9', 'hex');

  it('computes correct OPc', () => {
    const opc = computeOPc(ki, op);
    expect(opc.toString('hex')).toBe('cd63cb71954a9f4e48a5994e37a02baf');
  });

  it('generates correct f1 (MAC-A)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    // AUTN = (SQN^AK) || AMF || MAC-A, so MAC-A = AUTN[8..15]
    const macA = vectors.autn.subarray(8, 16).toString('hex');
    expect(macA).toBe('4a9ffac354dfafb3');
  });

  it('generates correct f2 (RES/XRES)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.xres.toString('hex')).toBe('a54211d5e3ba50bf');
  });

  it('generates correct f3 (CK)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.ck.toString('hex')).toBe('b40ba9a3c58b2a05bbf0d987b21bf8cb');
  });

  it('generates correct f4 (IK)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.ik.toString('hex')).toBe('f769bcd751044604127672711c6d3441');
  });

  it('generates correct f5 (AK)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.ak.toString('hex')).toBe('aa689c648370');
  });
});

// 3GPP TS 35.207 Test Set 2
describe('MILENAGE Test Set 2 (TS 35.207)', () => {
  const ki = Buffer.from('0396eb317b6d1c36f19c1c84cd6ffd16', 'hex');
  const op = Buffer.from('ff53bade17df5d4e793073ce9d7579fa', 'hex');
  const rand = Buffer.from('c00d603103dcee52c4478119494202e8', 'hex');
  const sqn = Buffer.from('fd8eef40df7d', 'hex');
  const amf = Buffer.from('af17', 'hex');

  it('computes correct OPc', () => {
    const opc = computeOPc(ki, op);
    expect(opc.toString('hex')).toBe('53c15671c60a4b731c55b4a441c0bde2');
  });

  it('generates correct f1 (MAC-A)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    const macA = vectors.autn.subarray(8, 16).toString('hex');
    expect(macA).toBe('5df5b31807e258b0');
  });

  it('generates correct f2 (RES/XRES)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.xres.toString('hex')).toBe('d3a628ed988620f0');
  });

  it('generates correct f3 (CK)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.ck.toString('hex')).toBe('58c433ff7a7082acd424220f2b67c556');
  });

  it('generates correct f4 (IK)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.ik.toString('hex')).toBe('21a8c1f929702adb3e738488b9f5c5da');
  });

  it('generates correct f5 (AK)', () => {
    const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
    expect(vectors.ak.toString('hex')).toBe('c47783995f72');
  });
});

// f5* and f1* tests (resync functions)
describe('MILENAGE Resync Functions (f5*, f1*)', () => {
  // Using Test Set 1 for deterministic values
  const ki = Buffer.from('465b5ce8b199b49faa5f0a2ee238a6bc', 'hex');
  const op = Buffer.from('cdc202d5123e20f62b6d676ac72cb318', 'hex');
  const rand = Buffer.from('23553cbe9637a89d218ae64dae47bf35', 'hex');

  describe('f5* (Anonymity Key for Resync)', () => {
    it('produces a 6-byte AK*', () => {
      const akStar = f5Star(ki, rand, op);
      expect(akStar.length).toBe(6);
    });

    it('produces AK* different from f5 AK (different constants)', () => {
      const vectors = generateVectorsWithRand(ki, op, rand, Buffer.from('ff9bb4d0b607', 'hex'), Buffer.from('b9b9', 'hex'));
      const ak = vectors.ak;
      const akStar = f5Star(ki, rand, op);
      expect(akStar.toString('hex')).not.toBe(ak.toString('hex'));
    });

    it('is deterministic for the same inputs', () => {
      const akStar1 = f5Star(ki, rand, op);
      const akStar2 = f5Star(ki, rand, op);
      expect(akStar1.toString('hex')).toBe(akStar2.toString('hex'));
    });
  });

  describe('f1* (MAC-S for Resync)', () => {
    it('produces an 8-byte MAC-S', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const amfZero = Buffer.alloc(2, 0);
      const macS = f1Star(ki, rand, sqnMs, amfZero, op);
      expect(macS.length).toBe(8);
    });

    it('produces MAC-S different from f1 MAC-A (different constants)', () => {
      const sqn = Buffer.from('ff9bb4d0b607', 'hex');
      const amf = Buffer.from('b9b9', 'hex');
      const amfZero = Buffer.alloc(2, 0);

      const vectors = generateVectorsWithRand(ki, op, rand, sqn, amf);
      const macA = vectors.autn.subarray(8, 16);
      const macS = f1Star(ki, rand, sqn, amfZero, op);

      // MAC-A uses different AMF and different constant C1, so they differ
      expect(macS.toString('hex')).not.toBe(macA.toString('hex'));
    });

    it('is deterministic for the same inputs', () => {
      const sqnMs = Buffer.from('000000000200', 'hex');
      const amfZero = Buffer.alloc(2, 0);
      const macS1 = f1Star(ki, rand, sqnMs, amfZero, op);
      const macS2 = f1Star(ki, rand, sqnMs, amfZero, op);
      expect(macS1.toString('hex')).toBe(macS2.toString('hex'));
    });
  });

  describe('validateAuts', () => {
    it('returns valid=true for correct AUTS', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);
      const result = validateAuts(ki, rand, auts, op);
      expect(result.valid).toBe(true);
      expect(result.sqnMs).toBeDefined();
      expect(result.sqnMs!.toString('hex')).toBe(sqnMs.toString('hex'));
    });

    it('returns valid=false for tampered AUTS (modified MAC-S)', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);
      // Tamper with the MAC-S (bytes 6-13)
      auts[10] ^= 0xff;
      const result = validateAuts(ki, rand, auts, op);
      expect(result.valid).toBe(false);
      expect(result.sqnMs).toBeUndefined();
    });

    it('returns valid=false for tampered AUTS (modified concealed-SQN)', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);
      // Tamper with the concealed-SQN (bytes 0-5)
      auts[3] ^= 0xff;
      const result = validateAuts(ki, rand, auts, op);
      expect(result.valid).toBe(false);
    });

    it('returns valid=false for wrong RAND', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);
      const wrongRand = Buffer.from('0000000000000000000000000000dead', 'hex');
      const result = validateAuts(ki, wrongRand, auts, op);
      expect(result.valid).toBe(false);
    });

    it('returns valid=false for wrong Ki', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);
      const wrongKi = Buffer.from('deadbeefdeadbeefdeadbeefdeadbeef', 'hex');
      const result = validateAuts(wrongKi, rand, auts, op);
      expect(result.valid).toBe(false);
    });

    it('returns valid=false for invalid AUTS length', () => {
      const result = validateAuts(ki, rand, Buffer.from('tooshort', 'utf8'), op);
      expect(result.valid).toBe(false);
    });

    it('round-trip: generate AUTS then validate extracts same SQN_MS', () => {
      // Test with various SQN values
      const testSqns = [
        Buffer.from('000000000000', 'hex'),
        Buffer.from('000000000001', 'hex'),
        Buffer.from('0000000000ff', 'hex'),
        Buffer.from('ffffffffffff', 'hex'),
        Buffer.from('123456789abc', 'hex'),
      ];

      for (const sqnMs of testSqns) {
        const auts = generateAuts(ki, rand, sqnMs, op);
        const result = validateAuts(ki, rand, auts, op);
        expect(result.valid).toBe(true);
        expect(result.sqnMs!.toString('hex')).toBe(sqnMs.toString('hex'));
      }
    });
  });

  describe('generateAuts', () => {
    it('produces a 14-byte AUTS', () => {
      const sqnMs = Buffer.from('000000000100', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);
      expect(auts.length).toBe(14);
    });

    it('AUTS structure: first 6 bytes concealed-SQN, next 8 bytes MAC-S', () => {
      const sqnMs = Buffer.from('abcdef012345', 'hex');
      const auts = generateAuts(ki, rand, sqnMs, op);

      // Extract parts
      const concealedSqn = auts.subarray(0, 6);
      const macS = auts.subarray(6, 14);

      // Verify concealed-SQN = SQN_MS XOR AK*
      const akStar = f5Star(ki, rand, op);
      const expectedConcealedSqn = Buffer.alloc(6);
      for (let i = 0; i < 6; i++) {
        expectedConcealedSqn[i] = sqnMs[i]! ^ akStar[i]!;
      }
      expect(concealedSqn.toString('hex')).toBe(expectedConcealedSqn.toString('hex'));

      // Verify MAC-S
      const amfZero = Buffer.alloc(2, 0);
      const expectedMacS = f1Star(ki, rand, sqnMs, amfZero, op);
      expect(macS.toString('hex')).toBe(expectedMacS.toString('hex'));
    });
  });
});

File: internal/crypto/milenage_test.go
package crypto

import (
	"encoding/hex"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func mustHex(s string) []byte {
	b, err := hex.DecodeString(s)
	if err != nil {
		panic(err)
	}
	return b
}

// 3GPP TS 35.207 Test Set 1
func TestMilenageTestSet1(t *testing.T) {
	ki := mustHex("465b5ce8b199b49faa5f0a2ee238a6bc")
	op := mustHex("cdc202d5123e20f62b6d676ac72cb318")
	rand := mustHex("23553cbe9637a89d218ae64dae47bf35")
	sqn := mustHex("ff9bb4d0b607")
	amf := mustHex("b9b9")

	t.Run("computes correct OPc", func(t *testing.T) {
		opc, err := ComputeOPc(ki, op)
		require.NoError(t, err)
		assert.Equal(t, "cd63cb71954a9f4e48a5994e37a02baf", hex.EncodeToString(opc))
	})

	t.Run("generates correct f1 (MAC-A)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		macA := hex.EncodeToString(vectors.AUTN[8:16])
		assert.Equal(t, "4a9ffac354dfafb3", macA)
	})

	t.Run("generates correct f2 (RES/XRES)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "a54211d5e3ba50bf", hex.EncodeToString(vectors.XRES))
	})

	t.Run("generates correct f3 (CK)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "b40ba9a3c58b2a05bbf0d987b21bf8cb", hex.EncodeToString(vectors.CK))
	})

	t.Run("generates correct f4 (IK)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "f769bcd751044604127672711c6d3441", hex.EncodeToString(vectors.IK))
	})

	t.Run("generates correct f5 (AK)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "aa689c648370", hex.EncodeToString(vectors.AK))
	})
}

// 3GPP TS 35.207 Test Set 2
func TestMilenageTestSet2(t *testing.T) {
	ki := mustHex("0396eb317b6d1c36f19c1c84cd6ffd16")
	op := mustHex("ff53bade17df5d4e793073ce9d7579fa")
	rand := mustHex("c00d603103dcee52c4478119494202e8")
	sqn := mustHex("fd8eef40df7d")
	amf := mustHex("af17")

	t.Run("computes correct OPc", func(t *testing.T) {
		opc, err := ComputeOPc(ki, op)
		require.NoError(t, err)
		assert.Equal(t, "53c15671c60a4b731c55b4a441c0bde2", hex.EncodeToString(opc))
	})

	t.Run("generates correct f1 (MAC-A)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		macA := hex.EncodeToString(vectors.AUTN[8:16])
		assert.Equal(t, "5df5b31807e258b0", macA)
	})

	t.Run("generates correct f2 (RES/XRES)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "d3a628ed988620f0", hex.EncodeToString(vectors.XRES))
	})

	t.Run("generates correct f3 (CK)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "58c433ff7a7082acd424220f2b67c556", hex.EncodeToString(vectors.CK))
	})

	t.Run("generates correct f4 (IK)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "21a8c1f929702adb3e738488b9f5c5da", hex.EncodeToString(vectors.IK))
	})

	t.Run("generates correct f5 (AK)", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		assert.Equal(t, "c47783995f72", hex.EncodeToString(vectors.AK))
	})
}

func TestMilenageResyncFunctions(t *testing.T) {
	ki := mustHex("465b5ce8b199b49faa5f0a2ee238a6bc")
	op := mustHex("cdc202d5123e20f62b6d676ac72cb318")
	rand := mustHex("23553cbe9637a89d218ae64dae47bf35")

	t.Run("f5* produces a 6-byte AK*", func(t *testing.T) {
		akStar, err := F5Star(ki, rand, op)
		require.NoError(t, err)
		assert.Len(t, akStar, 6)
	})

	t.Run("f5* produces AK* different from f5 AK", func(t *testing.T) {
		vectors, err := GenerateVectorsWithRAND(ki, op, rand, mustHex("ff9bb4d0b607"), mustHex("b9b9"))
		require.NoError(t, err)
		akStar, err := F5Star(ki, rand, op)
		require.NoError(t, err)
		assert.NotEqual(t, hex.EncodeToString(vectors.AK), hex.EncodeToString(akStar))
	})

	t.Run("f5* is deterministic", func(t *testing.T) {
		akStar1, err := F5Star(ki, rand, op)
		require.NoError(t, err)
		akStar2, err := F5Star(ki, rand, op)
		require.NoError(t, err)
		assert.Equal(t, hex.EncodeToString(akStar1), hex.EncodeToString(akStar2))
	})

	t.Run("f1* produces an 8-byte MAC-S", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		amfZero := make([]byte, 2)
		macS, err := F1Star(ki, rand, sqnMs, amfZero, op)
		require.NoError(t, err)
		assert.Len(t, macS, 8)
	})

	t.Run("f1* produces MAC-S different from f1 MAC-A", func(t *testing.T) {
		sqn := mustHex("ff9bb4d0b607")
		amf := mustHex("b9b9")
		amfZero := make([]byte, 2)

		vectors, err := GenerateVectorsWithRAND(ki, op, rand, sqn, amf)
		require.NoError(t, err)
		macA := hex.EncodeToString(vectors.AUTN[8:16])
		macS, err := F1Star(ki, rand, sqn, amfZero, op)
		require.NoError(t, err)

		assert.NotEqual(t, macA, hex.EncodeToString(macS))
	})

	t.Run("f1* is deterministic", func(t *testing.T) {
		sqnMs := mustHex("000000000200")
		amfZero := make([]byte, 2)
		macS1, err := F1Star(ki, rand, sqnMs, amfZero, op)
		require.NoError(t, err)
		macS2, err := F1Star(ki, rand, sqnMs, amfZero, op)
		require.NoError(t, err)
		assert.Equal(t, hex.EncodeToString(macS1), hex.EncodeToString(macS2))
	})
}

func TestValidateAUTS(t *testing.T) {
	ki := mustHex("465b5ce8b199b49faa5f0a2ee238a6bc")
	op := mustHex("cdc202d5123e20f62b6d676ac72cb318")
	rand := mustHex("23553cbe9637a89d218ae64dae47bf35")

	t.Run("returns valid=true for correct AUTS", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)
		result, err := ValidateAUTS(ki, rand, auts, op)
		require.NoError(t, err)
		assert.True(t, result.Valid)
		require.NotNil(t, result.SqnMs)
		assert.Equal(t, hex.EncodeToString(sqnMs), hex.EncodeToString(result.SqnMs))
	})

	t.Run("returns valid=false for tampered AUTS (modified MAC-S)", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)
		auts[10] ^= 0xff
		result, err := ValidateAUTS(ki, rand, auts, op)
		require.NoError(t, err)
		assert.False(t, result.Valid)
		assert.Nil(t, result.SqnMs)
	})

	t.Run("returns valid=false for tampered AUTS (modified concealed-SQN)", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)
		auts[3] ^= 0xff
		result, err := ValidateAUTS(ki, rand, auts, op)
		require.NoError(t, err)
		assert.False(t, result.Valid)
	})

	t.Run("returns valid=false for wrong RAND", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)
		wrongRand := mustHex("0000000000000000000000000000dead")
		result, err := ValidateAUTS(ki, wrongRand, auts, op)
		require.NoError(t, err)
		assert.False(t, result.Valid)
	})

	t.Run("returns valid=false for wrong Ki", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)
		wrongKi := mustHex("deadbeefdeadbeefdeadbeefdeadbeef")
		result, err := ValidateAUTS(wrongKi, rand, auts, op)
		require.NoError(t, err)
		assert.False(t, result.Valid)
	})

	t.Run("returns valid=false for invalid AUTS length", func(t *testing.T) {
		result, err := ValidateAUTS(ki, rand, []byte("tooshort"), op)
		require.NoError(t, err)
		assert.False(t, result.Valid)
	})

	t.Run("round-trip: generate AUTS then validate extracts same SQN_MS", func(t *testing.T) {
		testSqns := []string{
			"000000000000",
			"000000000001",
			"0000000000ff",
			"ffffffffffff",
			"123456789abc",
		}
		for _, sqnHex := range testSqns {
			sqnMs := mustHex(sqnHex)
			auts, err := GenerateAUTS(ki, rand, sqnMs, op)
			require.NoError(t, err)
			result, err := ValidateAUTS(ki, rand, auts, op)
			require.NoError(t, err)
			assert.True(t, result.Valid, "SQN: %s", sqnHex)
			assert.Equal(t, sqnHex, hex.EncodeToString(result.SqnMs))
		}
	})
}

func TestGenerateAUTS(t *testing.T) {
	ki := mustHex("465b5ce8b199b49faa5f0a2ee238a6bc")
	op := mustHex("cdc202d5123e20f62b6d676ac72cb318")
	rand := mustHex("23553cbe9637a89d218ae64dae47bf35")

	t.Run("produces a 14-byte AUTS", func(t *testing.T) {
		sqnMs := mustHex("000000000100")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)
		assert.Len(t, auts, 14)
	})

	t.Run("AUTS structure: first 6 bytes concealed-SQN, next 8 bytes MAC-S", func(t *testing.T) {
		sqnMs := mustHex("abcdef012345")
		auts, err := GenerateAUTS(ki, rand, sqnMs, op)
		require.NoError(t, err)

		concealedSqn := auts[:6]
		macS := auts[6:14]

		// Verify concealed-SQN = SQN_MS XOR AK*
		akStar, err := F5Star(ki, rand, op)
		require.NoError(t, err)
		expectedConcealedSqn := XOR(sqnMs, akStar)
		assert.Equal(t, hex.EncodeToString(expectedConcealedSqn), hex.EncodeToString(concealedSqn))

		// Verify MAC-S
		amfZero := make([]byte, 2)
		expectedMacS, err := F1Star(ki, rand, sqnMs, amfZero, op)
		require.NoError(t, err)
		assert.Equal(t, hex.EncodeToString(expectedMacS), hex.EncodeToString(macS))
	})
}
